codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/traits/ReferenceTraitValidator.java

```diff
@@ -0,0 +1,79 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.amazon.polymorph.traits;
+
+import java.util.ArrayList;
+import java.util.List;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.OperationIndex;
+import software.amazon.smithy.model.selector.Selector;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.Severity;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.smoketests.traits.SmokeTestCase;
+import software.amazon.smithy.smoketests.traits.SmokeTestsTrait;
+
+/**
+ * Validates correct usage of the @references trait,
+ * which is a type refinement trait and substantially changes
+ * how the shape is used in the model.
+ *
+ * This is not yet complete, especially since there are existing
+ * models out there misusing the trait.
+ * For now, we're only validating that smokeTests.params values
+ * never try to bind a node value to a reference shape,
+ * since there's no way to express a reference as a literal node value.
+ */
+public class ReferenceTraitValidator extends AbstractValidator {
```

Comment:
Blocking: TODO logic has continually bitten us in the ass (see .NET constraints issue).

Can you just scope this class down?

i.e: Call it what it does right now?
```suggestion
 * For now, we're only validating that smokeTests.params values
 * never try to bind a node value to a reference shape,
 * since there's no way to express a reference as a literal node value.
 * We could expand this to a proper {@code ReferenceTraitValidator} in the future.
 */
public class ReferenceTraitNotInSmokeTestValidator extends AbstractValidator {
```

Reply:
I think this is fair. It's not really customer-facing since having more validation error show up later with the validation event id of "ReferenceTrait" is reasonable. But I can see that the previous name gives implementors a false sense of confidence.

We do have https://github.com/smithy-lang/smithy-dafny/issues/315 to track the gap.